File: 2026/Elevator/01_Elevator_FCFS.py
from collections import deque
import asyncio, time

class Elevator:
    def __init__(self):
        self.current_floor = 0 # Ground floor

    def moveTo(self, floor):
        while self.current_floor != floor:
            if self.current_floor < floor:
                self.current_floor += 1
            else:
                self.current_floor -= 1
            time.sleep(1)  # Simulate time taken to move one floor
            print(f"[Elevator] Floor: {self.current_floor}")


class ElevatorSystem:
    def __init__(self):
        self.elevator = Elevator()
        self.requests = deque()
        self.isProcessing = False

    async def externalRequest(self, floor):
        print(f"[System] Call from Floor: {floor}")
        self.requests.append(floor)
        # print(f"Queue length: {len(self.requests)}")
        if not self.isProcessing:
            asyncio.create_task(self.processRequests())

    async def processRequests(self):
        self.isProcessing = True
        while self.requests:
            pickup = self.requests.popleft()
            print(f"--- Pickup at Floor: {pickup} ---")
            self.elevator.moveTo(pickup)

            dest = int(input("Enter destination: "))
            print(f"--- Dropoff at Floor: {dest} ---")
            self.elevator.moveTo(dest)
        self.isProcessing = False  

async def main():
    system = ElevatorSystem()
    await system.externalRequest(3)
    await system.externalRequest(7)
    await system.externalRequest(2)

if __name__ == "__main__":
    asyncio.run(main())

File: 2026/Elevator/01_Elevator_FCFS.cpp
#include<bits/stdc++.h>
using namespace std;

class Elevator{
public:
    int curr_floor = 0;

    void moveTo(int floor){
        while(curr_floor != floor){
            curr_floor += (curr_floor < floor) ? 1 : -1;
            this_thread::sleep_for(chrono::milliseconds(1000));
            cout<<"[Elevator] Floor: "<<curr_floor<<endl;
        }
    }
};

class ElevatorSystem{
    Elevator elevator;
    int requested_floor = -1;

public:
    void externalRequest(int floor){
        cout<<"[System] Call from Floor: "<<floor<<endl;
        requested_floor = floor;
        processRequests();
    }

    void processRequests(){
        int pickup = requested_floor;
        cout<<"--- Pickup at Floor: "<<pickup<<" ---"<<endl;
        elevator.moveTo(pickup);
         
        int dest;
        cout<<"Enter destination: "; cin>>dest;
        cout<<"--- Dropoff at Floor: "<<dest<<" ---"<<endl;
        elevator.moveTo(dest);
    }
};

int main(){
    ElevatorSystem system;
    cout<<"ELEVATOR STARTED AT FLOOR 0.\n\n";
    system.externalRequest(3);
    system.externalRequest(5);
}